data.md

```diff
@@ -200,5 +216,58 @@ Operator `_++_` implements an append operator for sort `Stack`.
 
     rule #drop(.ValTypes,   STACK)                       => STACK
     rule #drop(TYPE VTYPES, < TYPE > VAL:Number : STACK) => #drop(VTYPES, STACK)
+```
+
+Byte Map
+--------
+
+Wasm memories are byte arrays, sized in pages of 65536 bytes, initialized to be all zero bytes.
+To avoid storing many zeros in what may be sparse memory, we implement memory as maps, and store only non-zero bytes.
+The maps store integers, but maintains the invariant that each stored integer is between 1 and 255, inclusive.
+`BM [ N := I ]` writes the integer `I` to memory as bytes (little-endian), starting at index `N`.
+
+```k
+    syntax Map ::= Map "[" Int ":=" Int "]" [function]
+ // --------------------------------------------------
+    rule BM [ IDX := 0  ] => BM
+    rule BM [ IDX := VAL] =>
+        // Don't store 0 bytes.
+        #if VAL modInt 256 =/=Int 0
+          #then BM [IDX <- VAL modInt 256 ]
+          #else BM [IDX <- undef          ]
```

Comment:
Hmm, I think splitting up these as different rules rather than splitting on the right hand side would be nicer

Reply:
The Java K prover will only apply a given function rule if it can show that we are fully subsumed by that case. If you promote the splitting to the top-level rule (as @MrChico suggests), it will avoid duplicating the existing map `BM` twice in memory if it cannot prove that only one case of the function applies.

Another option is to fold the `#if_#then_#else_#fi` into the update itself:

```
...   => BM [ IDX <- #if VAL modInt 256 =/=Int 0 #then VAL modInt 256 #else undef ] ...
```

This avoids the memory explosion that you'll get with duplicating `BM` on the RHS when you can't prove that either `VAL modInt 256 =/=Int 0` or the negation is true, so would be another solution to the problem.

But I think @MrChico's solution is the best, because the term `BM [ IDX := VAL ]` held in memory is much smaller and easier to write lemmas about than the case where we evaluate the map fully.

Reply:
Thanks for the explanation, done.